src/Lean/Linter/UnusedVariables.lean

```diff
@@ -102,6 +102,11 @@ builtin_initialize addBuiltinUnusedVariablesIgnoreFn (fun _ stack opts =>
     (stx.isOfKind ``Lean.Parser.Term.matchAlt && pos == 1) ||
     (stx.isOfKind ``Lean.Parser.Tactic.inductionAltLHS && pos == 2))
 
+-- is anonymous have variable
+builtin_initialize addBuiltinUnusedVariablesIgnoreFn (fun stx stack _ =>
+    stx.getId == `this &&
+    (stack.matches [none, ``Lean.Parser.Term.haveIdDecl] ||
+    stack.matches [none, ``Lean.Parser.Term.haveEqnsDecl]))
```

Comment:
I still think that the unused variables linter should only look at original identifiers. More specifically, it uses `findReferences`, which was definitely not intended to return information not from the source text. If we ensure that, this workaround should not be necessary anymore, no?

Reply:
That sounds like a separate PR. I think the implementation of `unusedVariables` is weird in a bunch of ways - I remember noticing while straightening out the import graph that this file has an unexpected dependency on the LSP files, because it's using a function that was designed for a different purpose inside the server.

Reply:
I've restored the behavior of `findReferences`, please take a look. Btw, I was not able to reproduce the `have` issue.

Reply:
`findReferences` should probably be moved, yeah, but it is the right function to use here; recall e.g. https://github.com/leanprover/lean4/pull/1396